### PR TITLE
GoogleStackdriverTarget - Added support for TraceId and SpanId for LogEntry

### LINKED ISSUE #5677: Google.Cloud.Logging.Nlog - Log correlation with traces
Hi!

**Is your feature request related to a problem? Please describe.**
I can't correlate stackdriver logs in the tracing panel inside the google console.
image

I think that's because of the lack of fields like **trace**, **traceId** mentioned here https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry

I tried to find in code place where LogEntry is filled and eg. field **trace** exist, but it's skipped at **GoogleStackdriverTarget** during object creation.

`          var logEntry = new LogEntry
            {
                Severity = s_levelMap[loggingEvent.Level],
                Timestamp = ConvertToTimestamp(loggingEvent.TimeStamp),
                LogName = _logName,
                Resource = _resource,
            };
`
And at other places, algorithm flow fills labels, resources section, and nothing else.

 **Describe the solution you'd like**
Giving the possibility to fill LogEntry fields required for tracing.

If it's possible to then sorry for asking. However, in this case, documentation should be updated.

### PR DESCRIPTION
**Pull request by snakefoot**
Resolves #5677 by supporting [Log correlation with traces](https://cloud.google.com/trace/docs/trace-log-integration)

One could use [System.DiagnosticSource.Activity](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.activity) with help from [NLog.DiagnosticSource](https://www.nuget.org/packages/NLog.DiagnosticSource):

```xml
  <extensions>
    <add assembly="Google.Cloud.Logging.NLog" />
    <add assembly="NLog.DiagnosticSource" />
  </extensions>

  <targets>
    <target name="stackdriver" xsi:type="GoogleStackdriver" projectId="PROJECT_ID" logId="LOG_ID" traceId="${activity:TraceId}" spanId="${activity:SpanId}">
    </target>
  </targets>
```

Alternative one could capture TraceId using [System.Diagnostics.Trace.CorrelationManager.ActivityId](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.correlationmanager.activityid) and Nlog [${activityid}](https://github.com/NLog/NLog/wiki/Trace-Activity-Id-Layout-Renderer).

### CODE REVIEW

**Comment by snakefoot** [comment]
@amanda-tarafa + @jskeet Ready for review

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs [hunk 1]
```diff
@@ -488,6 +491,17 @@ private LogEntry BuildLogEntry(LogEventInfo loggingEvent)
                 Resource = _resource,
             };
 
+            if (!string.IsNullOrEmpty(_projectTraceId))
+            {
+                var traceId = RenderLogEvent(TraceId, loggingEvent);
+                if (!string.IsNullOrEmpty(traceId))
```
Comment (jskeet) [inline apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs:497]: Please add braces to the bodies of these single-statement `if` statements.

**Review by jskeet — COMMENTED** [review]
Assigned to Amanda who knows more about this area than I do. One formatting comment before then though.

**Comment by amanda-tarafa** [comment]
Thanks, I'll try to take a look today, else first thing Monday.

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs [hunk 2]
```diff
@@ -58,6 +58,7 @@ public partial class GoogleStackdriverTarget : TargetWithContext
         private Platform _platform;
         private MonitoredResource _resource;
         private string _logName;
+        private string _projectTraceId;
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs:61]:
```suggestion
        private string _traceResourcePath;
```

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs (hunk 1 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs:497]: +1 and everywhere else

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget_Configuration.cs [hunk 3]
```diff
@@ -98,6 +98,22 @@ public partial class GoogleStackdriverTarget
...
+        public Layout TraceId { get; set; }
+
+        /// <summary>
+        /// Activity SpanId for correlation of LogEvents from same activity.
+        /// </summary>
+        /// <remarks>
+        /// Trace API v2 uses: a 16-character hexadecimal encoding of an 8-byte array. Ex: `000000000000004a` = String.format("{0:x16}", 74)
+        /// </remarks>
+        public Layout SpanId { get; set; }
+
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget_Configuration.cs:116]: I think we'd want to add TraceSampled as well here, that should be true or false whether the trace has been sampled or not. And then we could add that value on the entry as well.

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs [hunk 4]
```diff
@@ -488,6 +491,17 @@ private LogEntry BuildLogEntry(LogEventInfo loggingEvent)
                 Resource = _resource,
             };
 
+            if (!string.IsNullOrEmpty(_projectTraceId))
+            {
+                var traceId = RenderLogEvent(TraceId, loggingEvent);
+                if (!string.IsNullOrEmpty(traceId))
+                    logEntry.Trace = _projectTraceId + traceId;
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs:498]:
```suggestion
                    logEntry.Trace = _projectTraceId + traceId;
                    logEntry.TraceSampled = // render and parse TraceSampled.
```

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]
Thanks a lot for this!
I've left a few change requests, and I think I'd also like to include an integration test.
I'm more than happy to take over the PR and make the changes myself, just leaving the review here in case you want to finish it yourself.

**Comment by snakefoot** [comment]
GitHub is currently broken, and makes it impossible for me to reply to review-comments.

But I have made commit that tries to resolves about review-comment.

**Comment by snakefoot** [comment]
GitHub is currently broken, and makes it impossible for me to reply to review-comments.

But I have made commit that tries to resolves about review-comment.

**Comment by snakefoot** [comment]
GitHub is currently broken, and makes it impossible for me to reply to review-comments.

But I have made commit that tries to resolves about review-comment.

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget_Configuration.cs (hunk 3 above)
Comment (orchowski) [reply apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget_Configuration.cs:116]: From my side, it can be done in a separate task, sampling is something additional, however, it'd be nice. Despite that, everything looks good to me.

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs [hunk 5]
```diff
@@ -491,86 +491,44 @@ private LogEntry BuildLogEntry(LogEventInfo loggingEvent)
...
-                        propertiesStruct.Fields.Add(combinedProperty.Key, jsonValue);
+                        logEntry.SpanId = spanId;
                     }
+
+                    logEntry.TraceSampled = TraceSampled;
                 }
-                logEntry.JsonPayload = jsonStruct;
+            }
+
+            if (SendJsonPayload)
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs:511]: Please undo the changes that are not related to #5677. It's totally fine if you want to refactor this code some, but please do so in another PR.

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget_Configuration.cs [hunk 6]
```diff
@@ -114,6 +114,13 @@ public partial class GoogleStackdriverTarget
         /// </remarks>
         public Layout SpanId { get; set; }
 
+        /// <summary>
+        /// True means that the trace resource name in the trace field was sampled for storage in a trace backend.
+        /// False means that the trace was not sampled for storage when this log entry was written, or the sampling decision was unknown at the time.
+        /// A non-sampled trace value is still useful as a request correlation identifier. The default is False.
+        /// </summary>
+        public bool TraceSampled { get; set; }
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget_Configuration.cs:122]: This value could be different per trace, it needs to be extracted from the log entry, it's not a configuration value.

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]
Some requested changes. Again, I'm happy to take over at any time. Thanks!

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget_Configuration.cs (hunk 6 above)
Comment (snakefoot) [reply apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget_Configuration.cs]: Changed to NLog Layout that accepts "true" or "1".

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs (hunk 5 above)
Comment (snakefoot) [reply apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs:515]: Undid refactoring

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs [hunk 7]
```diff
@@ -334,6 +336,7 @@ private void ActivateLogIdAndResource(string logId)
             var logName = new LogName(targetProjectId, logId);
             _logName = logName.ToString();
             _logNameToWrite = logName;
+            _traceResourcePath = !string.IsNullOrEmpty(targetProjectId) ? $"projects/{targetProjectId}/traces/" : string.Empty;
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs]: Changed from string.Empty to null.

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog.Tests/GoogleStackdriverTargetTest.cs [hunk 8]
```diff
@@ -279,6 +279,27 @@ public async Task SingleLogEntry()
...
+        {
+            var guidTraceId = Guid.NewGuid();
+
+            var uploadedEntries = await RunTestWorkingServer(
+                googleTarget =>
+                {
+                    googleTarget.SpanId = "${event-properties:SpanId:Format=x16}";
+                    googleTarget.TraceId = "${activityid}";
+                    googleTarget.TraceSampled = true;
+                    System.Diagnostics.Trace.CorrelationManager.ActivityId = guidTraceId;
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog.Tests/GoogleStackdriverTargetTest.cs:293]: Yes the NLog Layout provided here, is able to capture the TraceId from `CorrelationManager.ActivityId`:

`googleTarget.TraceId = "${activityid}";`

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog.Tests/GoogleStackdriverTargetTest.cs [hunk 9]
```diff
@@ -279,6 +279,27 @@ public async Task SingleLogEntry()
...
+                    LogManager.GetLogger("testlogger").Info("Hello {SpanId}", 74);
+                    return Task.FromResult(0);
+                });
+            Assert.Single(uploadedEntries);
+            Assert.Equal("000000000000004a", uploadedEntries[0].SpanId);
+            Assert.Equal($"projects/projectId/traces/{guidTraceId}", uploadedEntries[0].Trace);
+            Assert.Equal(true, uploadedEntries[0].TraceSampled);
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog.Tests/GoogleStackdriverTargetTest.cs]: Changed to `Assert.True`

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog.Tests/GoogleStackdriverTargetTest.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog.Tests/GoogleStackdriverTargetTest.cs:293]: Ack

**Review by amanda-tarafa — APPROVED** [review]
LGTM.

Thanks for this @snakefoot . I will merge now and add documentation later this week before releasing.

**Comment by snakefoot** [comment]
Created OR #6143 with the suggested refactoring.
